src/providers/aws.rs

```diff
@@ -1282,6 +1285,11 @@ impl RegionLauncher {
                     continue;
                 } else {
                     tracing::warn!("failed to terminate tsunami instances: {}", e);
+                    let e = Err(e).context("failed to terminate tsunami instances");
```

Comment:
If we fail to terminate the instances, I think we can return early with `?` — the operations below will almost certainly fail.

Reply:
The key stuff might still work?

Reply:
Imagine an example where the ec2 api is not working/failed/whatever, but a security group and key pair were still created. This way we clean those up.

Reply:
Except those cannot be deleted if the instances haven't been terminated, due to the dependency between them. Basically, they'll both fail with the dependency violation error.

Reply:
I don't think that is true for key pairs - they can definitely be deleted even if an instance using that key pair is still active. Also, deleting the key pair only deletes Amazon's copy of the public key. The instance has its own separate copy of the public key. So key pair deletion is independent of instance termination.

Reply:
Ah, interesting... How about we move up key deletion then? That way, we can then fail here if terminate fails, and the key is deleted regardless.

Reply:
Ah, but what then happens if key pair deletion fails? It's still possible that deleting the instances will work in that case, so that shouldn't short-circuit either?

Reply:
Yes, I think we move key deletion up, and if it errors, then we just log and continue. If termination errors, then we return an error.